README.md
# NAME

WebService::Whistle::Pet::Tracker::API - Perl interface to access the Whistle Pet Tracker Web Service

# SYNOPSIS

    use WebService::Whistle::Pet::Tracker::API;
    my $ws   = WebService::Whistle::Pet::Tracker::API->new(email=>$email, password=>$password);
    my $pets = $ws->pets; #isa ARRAY of HASHes
    foreach my $pet (@$pets) {
      print Dumper($pet);
    }

# DESCRIPTION

Perl interface to access the Whistle Pet Tracker Web Service.  All methods return JSON payloads that are converted to Perl data structures.  Methods that require authentication will request a token and cache it for the duration of the object.

# CONSTRUCTORS

## new

     my $ws = WebService::Whistle::Pet::Tracker::API->new(email=>$email, password=>$password);
    

# PROPERTIES

## email

Sets and returns the registered whistle account email

## password

Sets and returns the registered whistle account password

# METHODS

## pets

Returns a list of pets as an array reference

    my $pets = $ws->pets;

## device

Returns device data for the given device id

    my $device        = $ws->device('WXX-ABC123');
    my $battery_level = $device->{'battery_level'}; #0-100 charge level

## pet\_dailies

Returns dailies for the given pet id

    my $pet_dailies = $ws->pet_dailies($pet_id);

## pet\_daily\_items

Returns the daily items for the given pet id and day number

    my $pet_daily_items = $ws->pet_daily_items($pet_id, $day_number);

## pet\_stats

Returns pet stats for the given pet id

    my $pet_stats = $ws->pet_stats(123456789);

## places

Returns registered places as an array reference

    my $places = $ws->places;

# METHODS (API)

## api

Returns the decoded JSON data from the given web service end point

    my $data = $ws->api('/end_point');

## login

Calls the login service, caches, and returns the response.

## auth\_token

Retrieves the authentication token from the login end point

# ACCESSORS

## ua

Returns an [HTTP::Tiny](https://metacpan.org/pod/HTTP::Tiny) web client user agent

# SEE ALSO

    - L<https://github.com/RobertD502/whistleaio> Python implementation
    - L<https://github.com/martzcodes/node-whistle> NodeJS implementation (old api)

# AUTHOR

Michael R. Davis

# COPYRIGHT AND LICENSE

MIT License

Copyright (c) 2023 Michael R. Davis


Changes
Revision history for Perl extension WebService::Whistle::Pet::Tracker::API.

0.04 2023-05-13
  - Updated lib/WebService/Whistle/Pet/Tracker/API.pm
    - Added JWT decode and exp cache limit

0.03 2023-05-02
  - Updated lib/WebService/Whistle/Pet/Tracker/API.pm
    - Moved debug logging from Data::Dumper to JSON::XS
    - Cleaned up some error messages
  - Added scripts/perl-WebService-Whistle-Pet-Tracker-API-pets.pl
  - Added scripts/perl-WebService-Whistle-Pet-Tracker-API-device.pl
  - Added MQTT systemd service files and rpm package
  - Cleaned up scripts/perl-WebService-Whistle-Pet-Tracker-API-mqtt.pl

0.02 2023-04-23
  - Added scripts/perl-WebService-Whistle-Pet-Tracker-API-mqtt.pl
  - Updated lib/WebService/Whistle/Pet/Tracker/API.pm to clean up error messages

0.01 2023-04-22
  - original version


Makefile.PL
use ExtUtils::MakeMaker;
WriteMakefile(
  NAME           => 'WebService::Whistle::Pet::Tracker::API',
  LICENSE        => 'mit',
  VERSION_FROM   => 'lib/WebService/Whistle/Pet/Tracker/API.pm',
  PREREQ_PM      => {
                     'JSON::XS'          => 0,
                     'HTTP::Tiny'        => 0,
                     'Tie::IxHash'       => 0,
                     'Net::MQTT::Simple' => 0,
                     'Time::HiRes'       => 0,
                     'Crypt::JWT'        => 0,
                    },
  ABSTRACT_FROM  => 'lib/WebService/Whistle/Pet/Tracker/API.pm',
  AUTHOR         => 'Michael R. Davis',
    EXE_FILES    => [
                    qw{
                       scripts/perl-WebService-Whistle-Pet-Tracker-API-mqtt.pl
                       scripts/perl-WebService-Whistle-Pet-Tracker-API-pets.pl
                       scripts/perl-WebService-Whistle-Pet-Tracker-API-device.pl
                      }
                    ],
  'META_MERGE' => {
    'resources' => {
      'repository' => {
        'web' => 'https://github.com/mrdvt92/perl-WebService-Whistle-Pet-Tracker-API.git',
        'url' => 'git@github.com:mrdvt92/perl-WebService-Whistle-Pet-Tracker-API.git',
        'type' => 'git'
      },
      'homepage' => 'https://github.com/mrdvt92/perl-WebService-Whistle-Pet-Tracker-API',
      'bugtracker' => {
        'web' => 'https://github.com/mrdvt92/perl-WebService-Whistle-Pet-Tracker-API/issues'
      }
    },
    'meta-spec' => {
      'version' => 2
    }
  },
);


scripts/perl-WebService-Whistle-Pet-Tracker-API-mqtt.pl
#!/usr/bin/perl
use strict;
use warnings;
use File::Basename qw{basename};
use Getopt::Std qw{getopts};
use Tie::IxHash qw{};
use JSON::XS qw{encode_json};
use WebService::Whistle::Pet::Tracker::API qw{};
require Net::MQTT::Simple; #skip import

my $basename = basename($0);
my $syntax   = "$basename -h MQTT_HOST -e WHISTLE_EMAIL -p WHISTLE_PASSWORD #or from environment\n";
my $opt      = {};

getopts('e:p:h:', $opt);

my $host     = $opt->{'h'} || $ENV{'MQTT_HOST'}        || 'mqtt';
my $email    = $opt->{'e'} || $ENV{'WHISTLE_EMAIL'}    or die($syntax);
my $password = $opt->{'p'} || $ENV{'WHISTLE_PASSWORD'} or die($syntax);

my $mqtt     = Net::MQTT::Simple->new($host);
my $ws       = WebService::Whistle::Pet::Tracker::API->new(email=>$email, password=>$password);

my $pets     = $ws->pets;

foreach my $pet (@$pets) {
  my $topic = "stat/whistle-pet-tracker/pet/". $pet->{'id'};

  tie my %location, 'Tie::IxHash', (
                                    lat => $pet->{'last_location'}->{'latitude'},
                                    lon => $pet->{'last_location'}->{'longitude'},
                                    t   => $pet->{'last_location'}->{'timestamp'},
                                   );

  $mqtt->publish("$topic/name"                  => $pet->{'name'}                       );
  $mqtt->publish("$topic/last_location"         => encode_json(\%location)              );
  $mqtt->publish("$topic/device/battery_level"  => $pet->{'device'}->{'battery_level'}  );
  $mqtt->publish("$topic/device/battery_status" => $pet->{'device'}->{'battery_status'} );
}

$mqtt->tick(0.1);
$mqtt->disconnect;

__END__

=head1 NAME

perl-WebService-Whistle-Pet-Tracker-API-mqtt.pl - Publish Whistle Pet Tracker pet data to MQTT

=head1 SYNOPSIS

  perl-WebService-Whistle-Pet-Tracker-API-mqtt.pl -h MQTT_HOST -e WHISTLE_EMAIL -p WHISTLE_PASSWORD 

or

  export MQTT_HOST=mqtt
  export WHISTLE_EMAIL=my_email@example.com
  export WHISTLE_PASSWORD=my_password
  perl-WebService-Whistle-Pet-Tracker-API-mqtt.pl

=head1 DESCRIPTION

perl-WebService-Whistle-Pet-Tracker-API-mqtt.pl is a command line utility which connects to the configured MQTT broker and publishes pet data from the Whistle Pet Tracker API.

Topics:

  stat/whistle-pet-tracker/pet/123456789/name Rocky
  stat/whistle-pet-tracker/pet/123456789/last_location {"lat":38.88955018465634,"lon":-77.03530017768242,"t":"2023-04-23T13:55:01Z"}
  stat/whistle-pet-tracker/pet/123456789/device/battery_level 78
  stat/whistle-pet-tracker/pet/123456789/device/battery_status on

=head1 OPTIONS

-h Specifies the MQTT broker host name to which to connect and publish (default: mqtt)

-e Specifies the Whistle account name to use for authentication

-p Specifies the Whistle account password to use for authentication

=cut


scripts/perl-WebService-Whistle-Pet-Tracker-API-pets.pl
#!/usr/bin/perl
use strict;
use warnings;
use JSON::XS;
use File::Basename qw{basename};
use Getopt::Std qw{getopts};
use WebService::Whistle::Pet::Tracker::API qw{};

my $basename = basename($0);
my $syntax   = "$basename -e WHISTLE_EMAIL -p WHISTLE_PASSWORD #or from environment\n";
my $opt      = {};

getopts('e:p:', $opt);

my $email    = $opt->{'e'} || $ENV{'WHISTLE_EMAIL'}    or die($syntax);
my $password = $opt->{'p'} || $ENV{'WHISTLE_PASSWORD'} or die($syntax);

my $ws       = WebService::Whistle::Pet::Tracker::API->new(email=>$email, password=>$password);
my $json     = JSON::XS->new->pretty;
my $pets     = $ws->pets;

print $json->encode($pets);

__END__

=head1 NAME

perl-WebService-Whistle-Pet-Tracker-API-pets.pl - Get Whistle Pet Tracker pet data

=head1 SYNOPSIS

  perl-WebService-Whistle-Pet-Tracker-API-pets.pl -e WHISTLE_EMAIL -p WHISTLE_PASSWORD 

or

  export WHISTLE_EMAIL=my_email@example.com
  export WHISTLE_PASSWORD=my_password
  perl-WebService-Whistle-Pet-Tracker-API-pets.pl

=head1 DESCRIPTION

perl-WebService-Whistle-Pet-Tracker-API-pets.pl is a command line utility which gets pet data from the Whistle Pet Tracker API.

=head1 OPTIONS

-e Specifies the Whistle account name to use for authentication

-p Specifies the Whistle account password to use for authentication

=cut


scripts/perl-WebService-Whistle-Pet-Tracker-API-device.pl
#!/usr/bin/perl
use strict;
use warnings;
use JSON::XS;
use File::Basename qw{basename};
use Getopt::Std qw{getopts};
use WebService::Whistle::Pet::Tracker::API qw{};

my $basename = basename($0);
my $syntax   = "$basename -e WHISTLE_EMAIL -p WHISTLE_PASSWORD device [device [...]]\n";
my $opt      = {};

getopts('e:p:', $opt);

my $email    = $opt->{'e'} || $ENV{'WHISTLE_EMAIL'}    or die($syntax);
my $password = $opt->{'p'} || $ENV{'WHISTLE_PASSWORD'} or die($syntax);
die($syntax) unless @ARGV;

my $ws       = WebService::Whistle::Pet::Tracker::API->new(email=>$email, password=>$password);
my $json     = JSON::XS->new->pretty;

foreach my $argv (@ARGV) {
  my $device = $ws->device($argv);
  print $json->encode($device);
}

__END__

=head1 NAME

perl-WebService-Whistle-Pet-Tracker-API-device.pl - Get Whistle Pet Tracker device data

=head1 SYNOPSIS

  perl-WebService-Whistle-Pet-Tracker-API-device.pl -e WHISTLE_EMAIL -p WHISTLE_PASSWORD 

or

  export WHISTLE_EMAIL=my_email@example.com
  export WHISTLE_PASSWORD=my_password
  perl-WebService-Whistle-Pet-Tracker-API-device.pl

=head1 DESCRIPTION

perl-WebService-Whistle-Pet-Tracker-API-device.pl is a command line utility which gets device data from the Whistle Pet Tracker API.

=head1 OPTIONS

-e Specifies the Whistle account name to use for authentication

-p Specifies the Whistle account password to use for authentication

=cut


lib/WebService/Whistle/Pet/Tracker/API.pm
package WebService::Whistle::Pet::Tracker::API;
use strict;
use warnings;
use JSON::XS qw{};
use HTTP::Tiny qw{};
use Crypt::JWT qw{};

our $VERSION = '0.04';
our $PACKAGE = __PACKAGE__;
our $API_URL = 'https://app.whistle.com/api';

=head1 NAME

WebService::Whistle::Pet::Tracker::API - Perl interface to access the Whistle Pet Tracker Web Service

=head1 SYNOPSIS

  use WebService::Whistle::Pet::Tracker::API;
  my $ws   = WebService::Whistle::Pet::Tracker::API->new(email=>$email, password=>$password);
  my $pets = $ws->pets; #isa ARRAY of HASHes
  foreach my $pet (@$pets) {
    print JSON::XS->new->pretty->encode($pet);
  }

=head1 DESCRIPTION

Perl interface to access the Whistle Pet Tracker Web Service.  All methods return JSON payloads that are converted to Perl data structures.  Methods that require authentication will request a token and cache it for the duration of the object.

=head1 CONSTRUCTORS

=head2 new

  my $ws = WebService::Whistle::Pet::Tracker::API->new(email=>$email, password=>$password);

=cut

sub new {
  my $this  = shift;
  my $class = ref($this) ? ref($this) : $this;
  my $self  = {};
  bless $self, $class;
  %$self    = @_ if @_;
  return $self;
}

=head1 PROPERTIES

=head2 email

Sets and returns the registered Whistle account email

=cut

sub email {
  my $self         = shift;
  $self->{'email'} = shift if @_;
  die('Error: Whistle API: email required') unless $self->{'email'};
  return $self->{'email'};
}

=head2 password

Sets and returns the registered Whistle account password

=cut

sub password {
  my $self            = shift;
  $self->{'password'} = shift if @_;
  die('Error: Whistle API: password required') unless $self->{'password'};
  return $self->{'password'};
}

=head1 METHODS

=head2 pets

Returns a list of pets as an array reference

  my $pets = $ws->pets;

=cut

sub pets {
  my $self = shift;
  return $self->api('/pets')->{'pets'};
}

=head2 device

Returns device data for the given device id as a hash reference

  my $device        = $ws->device('WXX-ABC123');
  my $battery_level = $device->{'battery_level'}; #0-100 charge level

=cut

sub device {
  my $self          = shift;
  my $serial_number = shift or die('Error: Whistle API: Device serial number required.');
  return $self->api("/devices/$serial_number")->{'device'};
}

=head2 pet_dailies

Returns a list of dailies for the given pet id as an array reference

  my $pet_dailies = $ws->pet_dailies($pet_id);

=cut

sub pet_dailies {
  my $self   = shift;
  my $pet_id = shift or die('Error: Whistle API: pet id required');
  return $self->api("/pets/$pet_id/dailies")->{'dailies'}; #https://app.whistle.com/api/pets/123456789/dailies
}

=head2 pet_daily_items

Returns a list of the daily items for the given pet id and day number as an array reference

  my $pet_daily_items = $ws->pet_daily_items($pet_id, $day_number);

=cut

sub pet_daily_items {
  my $self       = shift;
  my $pet_id     = shift or die('Error: Whistle API: pet id required');
  my $day_number = shift or die('Error: Whistle API: day number required');
  return $self->api("/pets/$pet_id/dailies/$day_number/daily_items")->{'daily_items'};
}

=head2 pet_stats

Returns pet stats for the given pet id as a hash reference

  my $pet_stats = $ws->pet_stats(123456789);

=cut

sub pet_stats {
  my $self   = shift;
  my $pet_id = shift or die('Error: Whistle API: pet id required');
  return $self->api("/pets/$pet_id/stats")->{'stats'};
}

=head2 places

Returns registered places as an array reference

  my $places = $ws->places;

=cut

sub places {shift->api('/places')}; #this api call returns an array instead of a hash like other calls

=head1 METHODS (API)

=head2 api

Returns the decoded JSON data from the given web service end point

  my $data = $ws->api('/end_point');

=cut

sub api {
  my $self             = shift;
  my $api_destination  = shift or die('Error: Whistle API: api destination required');
  my $url              = $API_URL. $api_destination;
  my $response         = $api_destination eq '/login'
                       ? $self->ua->post_form($url, [email => $self->email, password => $self->password])
                       : $self->ua->get($url, {
                                               headers => {
                                                           Accept        => 'application/vnd.whistle.com.v6+json',
                                                           Authorization => 'Bearer '. $self->auth_token,
                                                          },
                                              }
                                       );
  print JSON::XS->new->pretty->encode({response => $response}) if $self->{'DEBUG'};
  my $status           = $response->{'status'};
  my $reason           = $response->{'reason'};
  if ($api_destination eq '/login') {
    die("Error: Whistle API: login failed\n") if $status eq 422;
    die("Error: Whistle API: request unsuccessful - request: $api_destination, status: $status $reason\n") unless $status eq 201;
  } else {
    die("Error: Whistle API: request unsuccessful - request: $api_destination, status: $status $reason\n") unless $status eq 200;
  }
  my $response_content = $response->{'content'};
  local $@;
  my $response_decoded = eval{JSON::XS::decode_json($response_content)};
  my $error            = $@;
  die("Error: Whistle API: invalid JSON - request: $api_destination, status: $status $reason, content: $response_content\n") if $error;
  print JSON::XS->new->pretty->encode({response_decoded => $response_decoded}) if $self->{'DEBUG'};
  return $response_decoded;
}

=head2 login

Calls the login service, caches, and returns the response as a hash reference.

=cut

sub login {
  my $self         = shift;
  $self->{'login'} = shift if @_;
  if (defined $self->{'login'}) {
    my $token   = $self->{'login'}->{'auth_token'};
    my $decoded = eval{Crypt::JWT::decode_jwt(token=>$token, ignore_signature=>1, verify_exp=>0)}; #{exp=>1685151718,user_id=>12345}
    if ($decoded and ref($decoded) eq 'HASH' and $decoded->{'exp'}) {
      delete($self->{'login'}) if time > $decoded->{'exp'};
    } else {
      delete($self->{'login'});
    }
  }
  $self->{'login'} = $self->api('/login') unless defined $self->{'login'};
  return $self->{'login'};
}

=head2 auth_token

Retrieves the authentication JWT token from the login end point

=cut

sub auth_token {shift->login->{'auth_token'}};

=head1 ACCESSORS

=head2 ua

Returns an L<HTTP::Tiny> web client user agent

=cut

sub ua {
  my $self = shift;
  unless ($self->{'ua'}) {
    my %settinges = (
                     keep_alive => 0,
                     agent      => "Mozilla/5.0 (compatible; $PACKAGE/$VERSION; See rt.cpan.org 35173)",
                    );
    $self->{'ua'} = HTTP::Tiny->new(%settinges);
  }
  return $self->{'ua'};
}

=head1 SEE ALSO

=over

=item Python - L<https://github.com/RobertD502/whistleaio>

=item NodeJS (old api) - L<https://github.com/martzcodes/node-whistle>

=back

=head1 AUTHOR

Michael R. Davis

=head1 COPYRIGHT AND LICENSE

MIT License

Copyright (c) 2023 Michael R. Davis

=cut

1;
